### PR TITLE
Task model take 2

### PR DESCRIPTION
**Pull request by LukeWood**
Same as https://github.com/keras-team/keras-cv/pull/1618 but with functional subclassing.

### CODE REVIEW

**Comment by LukeWood** [comment]
Ok so it looks like functional subclassing works.  Let me finish training a model, then we should merge.

File: keras_cv/models/object_detection/retina_net/retina_net_test.py [hunk 1]
```diff
@@ -199,3 +195,21 @@ def test_serialization(self):
         )
         serialized_2 = keras.utils.serialize_keras_object(restored)
         self.assertEqual(serialized_1, serialized_2)
+
+
+@pytest.mark.large
+class RetinaNetSmokeTest(tf.test.TestCase):
+    def test_backbone_preset_weight_loading(self):
+        # Check that backbone preset weights loaded correctly
+        _ = keras_cv.models.RetinaNet.from_preset(
```
Comment (jbischof) [inline keras_cv/models/object_detection/retina_net/retina_net_test.py:204]: Don't assign the input at all if unused

**Review by jbischof — COMMENTED** [review]
Thanks just some small comments!

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 2]
```diff
@@ -196,22 +223,43 @@ def __init__(
...
+        self.box_head = box_head
+        self.build(backbone.input_shape)
 
-        self.classification_head = (
-            classification_head
-            or layers_lib.PredictionHead(
-                output_filters=9 * num_classes,
-                bias_initializer=prior_probability,
+    @staticmethod
+    def build_forward(
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/retina_net/retina_net.py:232]: This feels a bit hacky to me, but maybe it's unavoidable?
If we do decide we want to go with this approach, I think this at least warrants a comment linking to a bug in Keras core describing the issue we're working around by doing this.

wdyt @LukeWood ?

**Review by ianstenbit — CHANGES_REQUESTED** [review]

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 2 above)
Comment (jbischof) [reply keras_cv/models/object_detection/retina_net/retina_net.py:232]: Why is it unavoidable? Sounds like I'm missing something

[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:232]: Interesting, I personally really dislike doing it in the constructor.  To me I feel it mixes up too many responsibilities/makes the constructor super chaotic.  Thoughts?

File: keras_cv/models/object_detection/retina_net/retina_net_test.py [hunk 3]
```diff
@@ -199,3 +195,21 @@ def test_serialization(self):
...
+            "resnet50_v2_imagenet",
+            num_classes=20,
+            bounding_box_format="xywh",
+        )
+
+    def test_full_preset_weight_loading(self):
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net_test.py:210]: Let’s wait until I have some weights I’m happy with locked in.

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:232]: Which part do you find hacky?  The functional subclassing?  The build_forwards method?

[the remark this answers is not in the document]
Comment (jbischof) [reply keras_cv/models/object_detection/retina_net/retina_net.py:232]: I think @fchollet's take would be that separating this out adds indirection without any payoff. We have inlined the functional model declaration across KerasNLP ([example](https://github.com/keras-team/keras-nlp/blob/0d7e4a91b94ce211085884f78fadaaeb6a8ef817/keras_nlp/models/bert/bert_backbone.py#L108)).

[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:232]: I see.  Not personally sold, but happy to compromise on this (at least in the short term to preserve consistency).  Will update this!

Comment (ianstenbit) [reply keras_cv/models/object_detection/retina_net/retina_net.py:232]: The part that I find hacky is that we have this separate method for building forward. The part that I meant was unavoidable was defining the forward pass in a method for construction time, but Jonathan points out that we can just do this in the constructor a la Functional model style.

That option SGTM (and bonus points that apparently NLP does that too!)

Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:232]: Seems we have consensus for doing it in line in the constructor.  Will update and make sure weights still load.

Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:232]: Fixed

**Comment by LukeWood** [comment]
Ok @jbischof and @ianstenbit, all comments addressed!

File: keras_cv/models/object_detection/retina_net/retina_net_test.py [hunk 4]
```diff
@@ -199,3 +195,25 @@ def test_serialization(self):
...
+        # Check that backbone preset weights loaded correctly
+        # TODO(lukewood): need to forward pass test once proper weights are
+        # implemented
+        _ = keras_cv.models.RetinaNet.from_preset(
```
Comment (jbischof) [inline keras_cv/models/object_detection/retina_net/retina_net_test.py:216]: Remove output assignment if unused

**Review by jbischof — APPROVED** [review]
Looks great! Some nits then ship away.

**Review by ianstenbit — APPROVED** [review]
thanks Luke!

**Comment by LukeWood** [comment]
😀

**Pull request merged**